src/tactic/apply_fun.lean

```diff
@@ -45,19 +45,75 @@ do {
+If the goal is of the form `a ≤ b` (or similarly for `a < b`), and `f` is an `order_iso`,
+we obtain a new goal `f a ≤ f b`.
+-/
+meta def apply_fun_to_goal (e : pexpr) (lem : option pexpr) : tactic unit :=
+do t ← target,
+  match t with
+  | `(%%l ≠ %%r) := (do
+      e' ← to_expr ``(ne_of_apply_ne %%e),
+      apply e',
+      skip)
+  | `(¬%%l = %%r) := (do
+      e' ← to_expr ``(ne_of_apply_ne %%e),
+      apply e',
+      skip)
+  | `(%%l = %%r) := (do
+      to_expr ``(%%e %%l), -- build and discard an application, to fill in implicit arguments
+      n ← get_unused_name `inj,
+      to_expr ``(function.injective %%e) >>= assert n,
+      -- Attempt to discharge the `injective f` goal
+      assumption <|>
+        (to_expr ``(equiv.injective) >>= apply >> skip) <|>
+        (lem.mmap (λ l, to_expr l >>= apply) >> skip),
+      -- Return to the main goal
+      swap,
+      n ← get_local n,
+      apply n,
+      clear n,
+      skip)
+  | `(%%l ≤ %%r) := (do
+      e' ← to_expr ``((order_iso.le_iff_le %%e).mp),
+      apply e',
+      skip)
+  | `(%%l < %%r) := (do
+      e' ← to_expr ``((order_iso.lt_iff_lt %%e).mp),
+      apply e',
+      skip)
+  | _ := fail ("failed to apply " ++ to_string e ++ " to the goal")
+  end
+
 namespace interactive
 
 setup_tactic_parser
 
 
 /--
-Apply a function to some local assumptions which are either equalities
-or inequalities. For instance, if the context contains `h : a = b` and
-some function `f` then `apply_fun f at h` turns `h` into
-`h : f a = f b`. When the assumption is an inequality `h : a ≤ b`, a side
-goal `monotone f` is created, unless this condition is provided using
-`apply_fun f at h using P` where `P : monotone f`, or the `mono` tactic
-can prove it.
+Apply a function to an equality or inequality in either a local hypothesis or the goal.
+
+* If we have `h : a = b`, then `apply_fun f at h` will replace this with `h : f a = f b`.
+* If we have `h : a ≤ b`, then `apply_fun f at h` will replace this with `h : f a ≤ f b`,
+  and create a subsidiary goal `monotone f`.
+  `apply_fun` will automatically attempt to discharge this subsidiary goal using `mono`,
+  or an explicit solution can be provided with `apply_fun f at h using P`, where `P : monotone f`.
+* If the goal is `a ≠ b`, `apply_fun f` will replace this with `f a ≠ f b`.
+* If the goal is `a = b`, `apply_fun f` will replace this with `f a = f b`,
+  and create a subsidiary goal `injective f`.
+  `apply_fun` will automatically attempt to discharge this subsidiary goal using local hypotheses,
+  or if `f` is actually an `equiv`,
+  or an explicit solution can be provided with `apply_fun f using P`, where `P : injective f`.
+* If the goal is `a ≤ b` (or similarly for `a < b`), and `f` is actually an `order_iso`,
+  `apply_fun f` will replace the goal with `f a ≤ f b`.
+  If `f` is anything else (e.g. just a function, or an `equiv`), `apply_fun` will fail.
+
```

Comment:
It seems the `using` optional part is not documented anywhere.

Reply:
Isn't it explained on lines 104 and 110?

Reply:
Could you add an example to the doc string showing the `using` feature?

Reply:
Sorry, I missed the explanation (that I probably wrote in the first place...).

Reply:
> Could you add an example to the doc string showing the `using` feature?

Could you do this below, after "Typical usage is"? My dream (which we're far from) is that the tactic docs show examples (not written explanations) of each feature of each tactic.